File: shoring/flywheel/src/main/java/com/panopset/flywheel/RawCommandLoader.java
package com.panopset.flywheel;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import com.panopset.compat.Logop;
import com.panopset.compat.Stringop;

class RawCommandLoader {

  private final Template tmplt;

  private final Deque<TemplateLine> queue = new ArrayDeque<>();

  private final List<Command> commands = new ArrayList<>();

  RawCommandLoader(final Template template) {
    assert template != null;
    tmplt = template;
  }

  List<Command> load() throws FlywheelException {
    tmplt.getTemplateSource().reset();
    int templateIndex = 0;
    int lineNumber = 0;
    while (!tmplt.getTemplateSource().isDone()) {
      if (tmplt.getFlywheel().isStopped()) {
        return new ArrayList<>();
      }
      flushQueue();
      if (tmplt.getFlywheel().isStopped()) {
        return new ArrayList<>();
      }
      String line = tmplt.getTemplateSource().next();
      if (tmplt.getTemplateRules().getLineBreaks().booleanValue()) {
        line = String.format("%s%s", line, Stringop.getEol());
      }
      process(new TemplateLine(line, templateIndex, lineNumber++));
      templateIndex += line.length();
    }
    flushQueue();
    return commands;
  }

  private void flushQueue() throws FlywheelException {
    while (!queue.isEmpty()) {
      process(queue.pop());
      if (tmplt.getFlywheel().isStopped()) {
        return;
      }
    }
  }

  private void loadCommand(final Command command) {
    if (tmplt.getFirstCommand() == null) {
      tmplt.setFirstCommand(command);
    }
    commands.add(command);
    Logop.debug("Loading command: " + command.toString());
  }

  private void process(final TemplateLine templateLine) throws FlywheelException {
    String line = templateLine.getLine();
    int openDirectiveLoc = line.indexOf(Syntax.getOpenDirective());
    int closeDirectiveLoc = line.indexOf(Syntax.getCloseDirective());
    if (closeDirectiveLoc == -1 || openDirectiveLoc == -1) {
      loadCommand(new CommandText(tmplt, templateLine, line));
      return;
    }
    if (closeDirectiveLoc < openDirectiveLoc) {
      skipTo(templateLine, openDirectiveLoc);
      return;
    }
    int endOfDirective = closeDirectiveLoc + Syntax.getCloseDirective().length();
    if (openDirectiveLoc == 0) {
      var command = new Command.Builder().template(tmplt).source(templateLine, closeDirectiveLoc).construct();
      if (command == null) {
        skipTo(templateLine, 1);
      } else {
        String remainder = line.substring(endOfDirective);
        loadCommand(command);
        if (remainder.length() > 0) {
          this.queue.push(new TemplateLine(remainder, templateLine.getTemplateCharIndex() + openDirectiveLoc, templateLine.getTemplateLineNumber()));
        }
      }
    } else {
      skipTo(templateLine, openDirectiveLoc);
    }
  }

  private void skipTo(final TemplateLine templateLine, final int pos) {
    String line = templateLine.getLine();
    int templateCharIndex = templateLine.getTemplateCharIndex();
    int templateLineNumber = templateLine.getTemplateLineNumber();
    String skippedTextLine = line.substring(0, pos);
    loadCommand(new CommandText(tmplt, new TemplateLine(skippedTextLine, templateCharIndex, templateLineNumber), skippedTextLine));
    this.queue.push(new TemplateLine(line.substring(pos), templateCharIndex + pos, templateLineNumber));
  }

  public static void addStructure(final List<Command> commands) {
    if (commands == null) {
      return;
    }
    Command prev = null;
    for (Command command : commands) {
      command.setPrev(prev);
      if (prev != null) {
        prev.setNext(command);
      }
      prev = command;
    }
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/RawCommandLoader.kt
package com.panopset.flywheel

import com.panopset.compat.Logz
import com.panopset.compat.Stringop
import java.util.*

internal class RawCommandLoader(val template: Template) {
    private val tmplt: Template = template
    private val queue: Deque<TemplateLine> = ArrayDeque()
    private val commands: MutableList<Command> = ArrayList()

    fun load(): List<Command> {
        tmplt.templateSource.reset()
        var templateIndex = 0
        var lineNumber = 0
        var firstTime = true
        while (!tmplt.templateSource.isDone) {
            if (tmplt.flywheel.isStopped) {
                return ArrayList()
            }
            flushQueue()
            if (tmplt.flywheel.isStopped) {
                return ArrayList()
            }
            var line = tmplt.templateSource.nextTemplateSourceLine()
            if (firstTime) {
                firstTime = false
            } else {
                if (tmplt.templateRules.lineBreaks) {
                    line = String.format("%s%s", Stringop.getEol(), line)
                }
            }
            process(TemplateLine(line, templateIndex, lineNumber++))
            templateIndex += line.length
        }
        flushQueue()
        return commands
    }

    private fun flushQueue() {
        while (!queue.isEmpty()) {
            process(queue.pop())
            if (tmplt.flywheel.isStopped) {
                return
            }
        }
    }

    private fun loadCommand(command: Command) {
        commands.add(command)
        Logz.debug("Loading command: $command")
    }

    private fun process(templateLine: TemplateLine) {
        val line = templateLine.line
        val openDirectiveLoc = line.indexOf(Syntax.getOpenDirective())
        val closeDirectiveLoc = line.indexOf(Syntax.getCloseDirective())
        if (closeDirectiveLoc == -1 || openDirectiveLoc == -1) {
            loadCommand(CommandText( tmplt, templateLine, line))
            return
        }
        if (closeDirectiveLoc < openDirectiveLoc) {
            skipTo(templateLine, openDirectiveLoc)
            return
        }
        val endOfDirective = closeDirectiveLoc + Syntax.getCloseDirective().length
        if (openDirectiveLoc == 0) {
            val command = CommandBuilder().template(tmplt).source(templateLine, closeDirectiveLoc).construct()
            if (command is CommandUnkown) {
                skipTo(templateLine, 1)
            } else {
                val remainder = line.substring(endOfDirective)
                loadCommand(command)
                if (remainder.isNotEmpty()) {
                    queue.push(
                        TemplateLine(
                            remainder,
                            templateLine.templateCharIndex + openDirectiveLoc,
                            templateLine.templateLineNumber
                        )
                    )
                }
            }
        } else {
            skipTo(templateLine, openDirectiveLoc)
        }
    }

    private fun skipTo(templateLine: TemplateLine, pos: Int) {
        val line = templateLine.line
        val templateCharIndex = templateLine.templateCharIndex
        val templateLineNumber = templateLine.templateLineNumber
        val skippedTextLine = line.substring(0, pos)
        loadCommand(
            CommandText( tmplt,
                TemplateLine(skippedTextLine, templateCharIndex, templateLineNumber),
                skippedTextLine
            )
        )
        queue.push(TemplateLine(line.substring(pos), templateCharIndex + pos, templateLineNumber))
    }
}

fun addStructure(commands: List<Command>) {
    var prev: Command? = null
    for (command in commands) {
        command.prev = prev
        if (prev != null) {
            prev.next = command
        }
        prev = command
    }
}
